Model export vignette

yulric opened the pull request:
This vignette documents a way that external investigators can export their model using CSV files. Look at the file `vignettes/model-export.Rmd`.

Things to discuss:
1. How to represent the hazards in a cox model?
2. How to represent the time metrics in a cox model? For eg., is the model for 1 year, 5 years, etc.

yulric committed:
First version of model export vignette

StaceyFisher commented:
1) center.csv - 'finalVariable' is the variable after centering, which may not actually be the final model variable, correct?
2) cox.csv - I assume that I can rename this file, depending on the type of model?
3) model-steps.csv - This file lists the order the files should be run, correct? So 'final' variables from the first file could be an input 'variable' in the second file- that would make sense to me
4) Should dummy variable specification be in another file? Or is it all specified in variable-details.csv? 
5) variable-details.csv - do the columns that are currently empty need to be filled? If so, could you describe what you want in them? I understand the ones that are currently filled
6) variable-details.csv - what variables go in this file? Just variables that need to be dummied? I see you have age as an example, however. Just age? Or also age_c, age_rcs1, and age_rcs2? I suppose maybe it is any variable that should be 'checked' (for min, max ect)? Perhaps it would help if we could specify at what stage this file is used, for example in model-steps.csv?
7) variables.csv - what variables go in this file? all starting variables?

yulric commented:
> 
> 
>     1. center.csv - 'finalVariable' is the variable after centering, which may not actually be the final model variable, correct?
> 
>     2. cox.csv - I assume that I can rename this file, depending on the type of model?
> 
>     3. model-steps.csv - This file lists the order the files should be run, correct? So 'final' variables from the first file could be an input 'variable' in the second file- that would make sense to me
> 
>     4. Should dummy variable specification be in another file? Or is it all specified in variable-details.csv?
> 
>     5. variable-details.csv - do the columns that are currently empty need to be filled? If so, could you describe what you want in them? I understand the ones that are currently filled
> 
>     6. variable-details.csv - what variables go in this file? Just variables that need to be dummied? I see you have age as an example, however. Just age? Or also age_c, age_rcs1, and age_rcs2? I suppose maybe it is any variable that should be 'checked' (for min, max ect)? Perhaps it would help if we could specify at what stage this file is used, for example in model-steps.csv?
> 
>     7. variables.csv - what variables go in this file? all starting variables?

1. Yes. Maybe it would be better to rename it to centeredVariable?
2. The fileName does not really matter but the step column value does. For eg., for a logistic regression model we would call the step logistic and the fileName could be model.csv. For the last step, the step column should inform us what kind of the model it is and so what formula needs to be run to evaluate the model.
3. Yes, the transformed variables from one could be inputs to the next file. Essentially, the variables that can be referenced in a file need to declared in some file before it.
4. I'm not sure. Do you think there's some dummying info that is not being captured in the variables-details file? The one good reason to put the dummy info there is because we are also defining the catValues there, so its easy to map a catValue to a dummy variable.
5. Ah sorry, those columns are part of the current variables-details file in cchsflow but I don't think we need them for the model export files. I'll remove them.
6. All the variables defined in the variables.csv file need to be defined in the variables-details file. 
7. All starting variables yes.

StaceyFisher commented:
> > ```
> > 1. center.csv - 'finalVariable' is the variable after centering, which may not actually be the final model variable, correct?
> > 
> > 2. cox.csv - I assume that I can rename this file, depending on the type of model?
> > 
> > 3. model-steps.csv - This file lists the order the files should be run, correct? So 'final' variables from the first file could be an input 'variable' in the second file- that would make sense to me
> > 
> > 4. Should dummy variable specification be in another file? Or is it all specified in variable-details.csv?
> > 
> > 5. variable-details.csv - do the columns that are currently empty need to be filled? If so, could you describe what you want in them? I understand the ones that are currently filled
> > 
> > 6. variable-details.csv - what variables go in this file? Just variables that need to be dummied? I see you have age as an example, however. Just age? Or also age_c, age_rcs1, and age_rcs2? I suppose maybe it is any variable that should be 'checked' (for min, max ect)? Perhaps it would help if we could specify at what stage this file is used, for example in model-steps.csv?
> > 
> > 7. variables.csv - what variables go in this file? all starting variables?
> > ```
> 
> 1. Yes. Maybe it would be better to rename it to centeredVariable?
> 2. The fileName does not really matter but the step column value does. For eg., for a logistic regression model we would call the step logistic and the fileName could be model.csv. For the last step, the step column should inform us what kind of the model it is and so what formula needs to be run to evaluate the model.
> 3. Yes, the transformed variables from one could be inputs to the next file. Essentially, the variables that can be referenced in a file need to declared in some file before it.
> 4. I'm not sure. Do you think there's some dummying info that is not being captured in the variables-details file? The one good reason to put the dummy info there is because we are also defining the catValues there, so its easy to map a catValue to a dummy variable.
> 5. Ah sorry, those columns are part of the current variables-details file in cchsflow but I don't think we need them for the model export files. I'll remove them.
> 6. All the variables defined in the variables.csv file need to be defined in the variables-details file.
> 7. All starting variables yes.

1. Yes, I would rename to centeredVariable
4. I think it is fine in the variable-details, as long as we know when dummy-ing occurs, defined in model-steps.csv

I will start to fill these in for DemPoRT.
Thanks!

yulric committed:
Updated the model exporting vignette

yulric committed:
Added interactions step to model-export vignette

StaceyFisher commented:
@yulric FYI - the vignette is inconsistent with the files. It says interactions are specified in the variable-details file and not in another file, however there it isn't in the example variable details file and there is another file. 
I would prefer the separate file approach

StaceyFisher commented:
Baseline hazard should go somewhere too. Not sure where would make the most sense. @yulric

yulric commented:
> 
> 
> Baseline hazard should go somewhere too. Not sure where would make the most sense. @yulric

I think in the cox step or in your case the fine-and-grey step. Should we have a row in that file for baseline hazard? I know that some algorithms the baseline hazard changes with time?

yulric committed:
Fixed bug in model export vignette due to wrong path to example interactions CSV file

DougManuel commented on vignettes/assets/model-export/variables.csv:
```diff
@@ -0,0 +1,5 @@
+variable,label,labelLong,section,subject,variableType,units,description
```
The example of variables.csv is the starting variables, not the variables that end up in the model will be transformed. The final variables will be sex_cat1_C_1.

DougManuel replied:
Keep in mind reference population

yulric replied:
Yes, the variables.csv files should document all the variables that are not part of any subsequent step in the algorithm. For eg., you would not put the RCS variables in here because they would be documented in the RCS file.

DougManuel commented on vignettes/assets/model-export/variables-details.csv:
```diff
@@ -0,0 +1,7 @@
+﻿variable,dummyVariable,numValidCat,catLabel,catLabelLong,catValue,units,type,interval,notes
+age,NA,NA,NA,NA,NA,years,cont,"[26,78]",
```
Keep in mind that max mins will be different from the original variable_details.csv sheet. 

I think we will need outlier treatment in variables or variable details. This method could be in model steps, since outliers is really a transformation.

DougManuel commented on vignettes/assets/model-export/center.csv:
```diff
@@ -0,0 +1,2 @@
+﻿variable,centerValue,centeredVariable
```
Suggest that this table look more like a Table 1.

Meaning,
- include some descriptions like median, 95CIs, IQR. 
- stratification will often be needed. eg. sex. 

So, this table be called descriptive.csv or something similar. We don't need to be too prescriptive about what the file contains - could vary by algorithm. 

It could be all or part of the file from RDF.

yulric replied:
This file is not a descriptive table for the variables but to document how to center certain variables in the algorithm.

CBjerke committed:
suggested edits to text

CBjerke committed:
Edits to text/format

CBjerke commented on vignettes/model-export.Rmd:
````diff
@@ -0,0 +1,191 @@
+
+Once we have our variables sheet, we can make our variables details sheet.
+
+```{r}
````
Suggest changing catLabel to "D-penicillamine" and keeping catLabelLong "Received D-penicillamine" to better illustrate the difference between details for these two columns

CBjerke commented on vignettes/model-export.Rmd:
````diff
@@ -0,0 +1,191 @@
+Once we have our variables sheet, we can make our variables details sheet.
+
+```{r}
+pbc_variables_details <- read.csv(file.path(getwd(), "/assets/model-export/variables-details.csv"))
````
I also think that the column should be named variableType (as in the variables sheet)

CBjerke commented on vignettes/model-export.Rmd:
```diff
@@ -0,0 +1,191 @@
+
+The cox.csv file gives us the names of the final variables in the model and their coefficients.
+
+There is no file for the dummy step since we already defined the dummy variables in the variables details file, giving us all the information we need to perform that step.
```
I think we should provide an example of what would be needed in this file. Perhaps it is better not to simplify the 'status variable' earlier on in the vignette so that details of what would be required could be included here.

CBjerke commented on vignettes/model-export.Rmd:
````diff
@@ -0,0 +1,191 @@
+DT::datatable(pbc_variables)
+```
+
+The variables sheet gives a summary of all the starting variables used in the model. A brief description of its columns are outlined below,
````
"Starting" variable is a little confusing, do you really mean this? For example, we would have 'packyears' in this column and not the start variables 'SMKDSTY', 'SMK_01A', etc. Or, do I have this wrong.

yulric replied:
Yes, thinking now I believe this file should include all the variables in the algorithm that are not part of any other step file.

CBjerke reviewed (changes requested):
@yulric I made a number of edits in the text in a new branch but I can't figure out, in R studio, how to push these changes for your review.

CBjerke committed:
edit

DougManuel reviewed (changes requested):
I really like how you've walked through an example of the whole algorithm development. 

A few overarching comments:
1) The **primary purpose of export** is to provide information to allow other to calculate risk using our API, their own implementation, etc. Right? I think we can have other documentation if people want to recreate the entire algorithm development process. Or the model export is part of the overall development to implementation to validation process.

2) **Model description** - We'll need overall model information: algorithm name, creator, etc. We originally had this information on a google sheet and then we've started to develop the config file. the config file is the way to go, but it has quite a bit of information that we would not need for model export. 

3) **variables.csv** the key information is, _I think_ final variables, not starting variables. For discussion. I think we may want to have three types of variables.
- final variables - the exact final variables that are used to estimate the algorithm. This would basically be a description of any variable that has a beta coefficient. There should be a one-to-one-match with beta coefficients. We have labels, etc. for every variable in the beta sheet.
- starting variables from our development data - I actually don't think this is as important, but we need for batch mode.
- starting variables for LIME or our web implementation. Again, not as important for many people, but I think helpful for ourselves and others.
- we probably don't need to include other implementations (say St. Vincent, for RESPECT) but this is a consideration. We probably want separate documentation for specific implementation in a setting like St. V.

4) **Metadata** - we need a file to describe the variables headers in this documentation definitions for such things as 'coefficient', rcsVariables, knots, centreVariable, etc. This could be a separate reference file that we use for multiple models. If so, this reference file needs to be in the model description file with a version of the metadata (since it will change). We've started this metadata in a few places, including google sheets and @DarkShadeKnigh R code within bllflow.

DougManuel commented:
I've added @Rhan43 to this PR review, since she'll need to create oversee the creation of model export for RESPECT. Also included @amytmhsu in case she want to see what's going on.

yulric committed:
Merge pull request #71 from Big-Life-Lab/model-export-edits

Model export edits

DougManuel committed:
[feature] add description.csv

DougManuel committed:
[feature] initial commit of metadata

From https://docs.google.com/spreadsheets/d/1jWXOdL2EX18xm-PB7syML1FJUNvWoLJFxNlkICFkjyU/edit#gid=1027182114

DougManuel committed:
[feature] update metadata used in different export files

DougManuel committed:
[feature] update metadata

CBjerke committed:
test commit

CBjerke committed:
[refactor] updates to variable_details plus meta-data

CBjerke committed:
[update] update description file

CBjerke committed:
[update] update variable-details

CBjerke committed:
[update] to rcs file

CBjerke committed:
[update] to rcs file

CBjerke committed:
[update] to metadata file

CBjerke committed:
[refactor] metadata formatted for machine readability

CBjerke committed:
[featrue] missing outlier csv template

CBjerke committed:
[refactor] add/remove fields from missing outlier csv file

Rhan43 reviewed (commented):
For discussion: we've listed files specific to this specific stage "describing a model". However, I'm still not 100% clear how these files fit relative to other files along the entire process: initial specification to web implementation. Ideally, we'd have one file to use at each stage its needed. 
As I mentioned in a comment variables and variable_detials are used to describe two different stages: "prespecifying your data" and "describing your prediction model", however the details required for each file is currently different for each stage.
Also a potential issue, is with "describing your prediction model" and "implementing to web". In "describing your prediction model" we ask for a center.csv, but in implementing to web, we also use the centering values to deal with errors (e.g., input value is missing).

CBjerke committed:
Merge pull request #72 from Big-Life-Lab/model-export-description

Initial specification for description.csv and metadata.csv

yulric commented:
Closing PR. All documentation is in the [model parameters repo](https://github.com/Big-Life-Lab/model-parameters).
